NebulaNetwork/Ngrok/NgrokManager.cs

```diff
@@ -0,0 +1,212 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Compression;
+using System.Net.Http;
+using System.Reflection;
+using System.Threading.Tasks;
+using NebulaModel;
+using NebulaModel.Utils;
+using NebulaWorld;
+
+namespace NebulaNetwork.Ngrok
+{
+    public class NgrokManager
+    {
+        private readonly string _ngrokPath = $"{Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)}/ngrok-v3-stable-windows-amd64/ngrok.exe";
+        private readonly int _port;
+        private readonly string _authtoken;
+
+        private Process _ngrokProcess;
+
+        public string NgrokAddress;
+
+        public NgrokManager(int port, string authtoken = null)
+        {
+            _port = port;
+            _authtoken = authtoken ?? Config.Options.NgrokAuthtoken;
+
+            if (!Config.Options.EnableNgrok)
+            {
+                return;
+            }
+
+            if (string.IsNullOrEmpty(_authtoken))
+            {
+                NebulaModel.Logger.Log.Warn("Ngrok support was enabled, however no Authtoken was provided");
+                return;
+            }
+
+            // Start this stuff in it's own thread, as we require async and we dont want to freeze up the GUI when freeze up when Downloading and installing ngrok
+            Task.Run(async () =>
+            {
+
+                if (!IsNgrokInstalled())
+                {
+                    var hasDownloadAndInstallBeenConfrimed = false;
+                    var downloadAndInstallConfrimation = new Task<bool>(() => true);
+                    var downloadAndInstallRejection = new Task<bool>(() => false);
+
+                    UnityDispatchQueue.RunOnMainThread(() =>
+                    {
+                        InGamePopup.ShowWarning(
+                            "Ngrok download and installation confrimation",
+                            "Ngrok is support is enabled, however it has not been downloaded and installed yet, do you want to automattically download and install Ngrok?",
+                            "Accept",
+                            "Reject",
+                            () => downloadAndInstallConfrimation.Start(),
+                            () => downloadAndInstallRejection.Start()
+                        );
+                    });
+
+                    hasDownloadAndInstallBeenConfrimed = await await Task.WhenAny(downloadAndInstallConfrimation, downloadAndInstallRejection);
+                    if (!hasDownloadAndInstallBeenConfrimed)
```

Comment:
```suggestion
                    var hasDownloadAndInstallBeenConfirmed = false;
                    var downloadAndInstallConfirmation = new Task<bool>(() => true);
                    var downloadAndInstallRejection = new Task<bool>(() => false);

                    UnityDispatchQueue.RunOnMainThread(() =>
                    {
                        InGamePopup.ShowWarning(
                            "Ngrok download and installation confirmation",
                            "Ngrok is support is enabled, however it has not been downloaded and installed yet, do you want to automattically download and install Ngrok?",
                            "Accept",
                            "Reject",
                            () => downloadAndInstallConfirmation.Start(),
                            () => downloadAndInstallRejection.Start()
                        );
                    });

                    hasDownloadAndInstallBeenConfirmed = await await Task.WhenAny(downloadAndInstallConfirmation, downloadAndInstallRejection);
                    if (!hasDownloadAndInstallBeenConfirmed)
```

Reply:
@PhantomGamers I am trying to find the actual change's in this suggestion :o however I can't seem to find them :o

Reply:
typo confrim -> confirm

Reply:
Ah ic :)